scripts/h5toms.py

```diff
@@ -335,78 +368,117 @@
-                weight_data = np.dstack(weight_pol_data)
-                flag_data = np.dstack(flag_pol_data)
-
-                model_data = None
-                corrected_data = None
-                if options.model_data:
-                    # unity intensity zero phase model data set, same shape as vis_data
-                    model_data = np.ones(vis_data.shape, dtype=np.complex64)
-                    # corrected data set copied from vis_data
-                    corrected_data = vis_data
-
-                uvw_coordinates = np.array(target.uvw(ant2, timestamp=out_utc, antenna=ant1))
-
-                # Convert averaged UTC timestamps to MJD seconds.
-                out_mjd = [katpoint.Timestamp(time_utc).to_mjd() * 24 * 60 * 60 for time_utc in out_utc]
-
-                # Increment the filesize.
-                sz_mb += vis_data.dtype.itemsize * vis_data.size / (1024.0 * 1024.0)
-                sz_mb += weight_data.dtype.itemsize * weight_data.size / (1024.0 * 1024.0)
-                sz_mb += flag_data.dtype.itemsize * flag_data.size / (1024.0 * 1024.0)
-
-                if options.model_data:
-                    sz_mb += model_data.dtype.itemsize * model_data.size / (1024.0 * 1024.0)
-                    sz_mb += corrected_data.dtype.itemsize * corrected_data.size / (1024.0 * 1024.0)
-
-                # write the data to the ms.
-                main_dict = ms_extra.populate_main_dict(uvw_coordinates, vis_data, flag_data,
-                                                        out_mjd, ant1_index, ant2_index,
-                                                        dump_time_width, field_id, state_id,
-                                                        scan_itr, model_data, corrected_data)
-                ms_extra.write_rows(main_table, main_dict, verbose=options.verbose)
+        # Iterate over time in some multiple of dump average
+        ntime = utc_seconds.size
+        tsize = dump_av*100
+        out_freqs = h5.channel_freqs
+        nchan = out_freqs.size
+
+        for ltime in xrange(0, ntime, tsize):
+            utime = min(ltime + tsize, ntime)
+            tdiff = utime - ltime
+
+            # load all visibility, weight and flag data
+            # for this scan's timestamps.
+            # Ordered (ntime, nchan, nbl*npol)
+            scan_data = h5.vis[ltime:utime, :, :]
+            scan_weight_data = h5.weights[ltime:utime, :, :]
+            scan_flag_data = h5.flags[ltime:utime, :, :]
+
+            # Select correlator products
+            # cp_index could be used above when the LazyIndexer
+            # supports advanced integer indices
+            vis_data = scan_data[:,:,cp_index]
+            weight_data = scan_weight_data[:,:,cp_index]
+            flag_data = scan_flag_data[:,:,cp_index]
+
+            # Zero any invalid correlator products
+            vis_data[:,:,invalid_cp] = 0 + 0j
+            weight_data[:,:,invalid_cp] = 0
+            flag_data[:,:,invalid_cp] = False
```

Comment:
@ludwigschwardt  If there's no correlator product available, should we not flag the data? However, the above reflects original functionality.